### PR TITLE
[RFC] Remove OneIndexedArray

### PR DESCRIPTION
**Pull request by joehuchette**
This replaces ``OneIndexedArray{T,N}`` with a plain ``Array{T,N}``, which simplifies a lot, especially in the vectorization code (in particular, the concatenation code is no longer needed!). To accomplish this, I changed the ``JuMPContainer`` interface: they now have only two fields, the data field, and a metadata dictionary that currently only has ``:model=>JuMP.Model()``. I moved all the indexing stuff used in the printing code into a ``ObjectIdDict`` stored in the model, which seems like a more appropriate place for it anyway. The only regression I notice is that I can't find out the model attached to an ``Array(Variable, 0)``, which seems like a sufficiently corner case.

This touches a surprising amount of code, so I'd appreciate reviews.

### CODE REVIEW

**Comment by joehuchette** [comment]
Oh, and I added ``m.varCones``, which right now is only used to record the SDP variables, but could be reused in the future for other cones as well.

**Comment by joehuchette** [comment]
(Also, ``m.dictList`` could probably replaced with ``keys(m.varData)``...)

File: src/JuMPContainer.jl [hunk 1]
```diff
@@ -52,68 +68,79 @@ Base.isempty(d::JuMPContainer) = isempty(_innercontainer(d))
...
-                offset[i] = 1 - idxsets[i].args[1]
-            else
-                error("Currently only ranges with integer compile-time starting values are allowed as index sets. $(idxsets[i].args[1]) is not an integer in range $(idxsets[i]).")
-            end
-        end
         truearray = all(idxsets) do rng
             isexpr(rng,:(:)) && # redundant, but might as well
             rng.args[1] == 1 && # starts from 1
-            length(rng.args) == 2 || rng.args[2] == 1 # steps of one
+            (length(rng.args) == 2 || rng.args[2] == 1) # steps of one
```
Comment (joehuchette) [inline src/JuMPContainer.jl]: Think this is a bug fix, as well

File: src/JuMPContainer.jl [hunk 2]
```diff
@@ -52,68 +68,79 @@ Base.isempty(d::JuMPContainer) = isempty(_innercontainer(d))
...
+                    error("Currently only ranges with integer compile-time starting values are allowed as index sets. $(idxsets[i].args[1]) is not an integer in range $(idxsets[i]).")
+                end
+            end
+            typecode = quote
+                type $(typename){T} <: JuMPArray{T,$N}
+                    innerArray::Array{T,$N}
+                    meta::Dict{Symbol,Any}
+                end
+            end
+            constrlhs = :($(typename)(innerArray::Array))
```
Comment (joehuchette) [inline src/JuMPContainer.jl]: Could maybe make this (and a lot of the other methods here?) a single method for ``JuMPArray``, outside the ``eval`` block.

File: src/JuMPContainer.jl [hunk 3]
```diff
@@ -208,16 +237,18 @@ function Base.next(x::JuMPArray,k)
     tuple(keys..., x[keys...]), gidx
 end
 
-_next_index{T,N}(x::JuMPArray{T,N}, k) =
+function _next_index{T,N}(x::JuMPArray{T,N}, k)
+    data = printdata(x)
```
Comment (joehuchette) [inline src/JuMPContainer.jl]: I guess the abstraction is pretty leaky here

File: src/JuMPContainer.jl [hunk 4]
```diff
@@ -239,6 +269,13 @@ Base.values(d::JuMPDict)  = values(d.tupledict)
 Base.keys(d::JuMPArray)   = KeyIterator(d)
 Base.values(d::JuMPArray) = ValueIterator(d.innerArray)
 
+# TODO: add keys/values for Array
+Base.keys(d::Array) = CartesianRange(size(d))
+Base.start(ci::CartesianIndex)    = start(ci.I)
+Base.next( ci::CartesianIndex, k) = next(ci.I, k)
+Base.done( ci::CartesianIndex, k) = done(ci.I, k)
+Base.values(d::Array) = d
```
Comment (joehuchette) [inline src/JuMPContainer.jl]: These methods would probably be more appropriate in ``Base``, if deemed appropriate.

File: test/sdp.jl [hunk 5]
```diff
@@ -155,7 +155,7 @@ context("With solver $(typeof(solver))") do
     XX, yy = getValue(X), getValue(y)
     @fact ispsd(XX) --> true
     @fact (yy[0] >= 0) --> true
-    @fact (yy[1]^2 + yy[2]^2 <= yy[0]^2 + 1e-5) --> true
+    @fact (yy[1]^2 + yy[2]^2 <= yy[0]^2 + 1e-4) --> true
```
Comment (joehuchette) [inline test/sdp.jl]: Should probably go on master, this (should be) unrelated

File: src/JuMPContainer.jl (hunk 4 above)
Comment (joehuchette) [inline src/JuMPContainer.jl]: They also aren't in v0.3, hence the Travis failures

File: src/JuMPContainer.jl [hunk 6]
```diff
@@ -17,19 +16,36 @@ end
...
+function JuMPDict{T,N}(d::Dict{NTuple{N,Any},T})
+    tmp = JuMPDict{T,N}()
+    tmp.tupledict = d
+    tmp
+end
+
+function JuMPDict{T,N}(d::Dict{NTuple{N,Any},T}, meta::Dict{Symbol,Any})
+    tmp = JuMPDict{T,N}()
+    tmp.tupledict = d
+    tmp.meta = meta
```
Comment (joehuchette) [inline src/JuMPContainer.jl]: Not ideal that we have to construct two unnecessary (empty) Dicts here, but I couldn't figure out how to make the inner constructors work like I wanted them to.

File: src/JuMPContainer.jl (hunk 2 above)
Comment (mlubin) [inline src/JuMPContainer.jl]: That's okay, this will all go away soon

**Comment by mlubin** [comment]
Any performance regressions on the benchmarks?

**Comment by joehuchette** [comment]
PR:
```
❯ j test/perf/macro.jl                                                    JuMP/git/O̶n̶e̶I̶n̶d̶e̶x̶e̶d̶A̶r̶r̶a̶y̶ !
  Running N=20...
    N=20 min 0.012108239
    N=20 min 0.025968998
  Running N=50...
    N=50 min 0.201408882
    N=50 min 0.460831409
  Running N=100...
    N=100 min 1.813099589
    N=100 min 3.956056257

❯ j test/perf/vector_speedtest.jl                                         JuMP/git/O̶n̶e̶I̶n̶d̶e̶x̶e̶d̶A̶r̶r̶a̶y̶ !
n = 10
Vector with sum{}: 1.1096e-5
Vector with dot() : 1.3291e-5
2D Matrix with sum{}: 5.762e-6
2D Matrix with bigdot(): 0.01451667
3D Matrix with sum{}: 5.7295e-5
3D Matrix with dot(): 0.000112291
n = 50
Vector with sum{}: 1.1415e-5
Vector with dot() : 1.3373e-5
2D Matrix with sum{}: 8.918e-5
2D Matrix with bigdot(): 0.000186671
3D Matrix with sum{}: 0.006058401
3D Matrix with dot(): 0.015342567
n = 100
Vector with sum{}: 1.3654e-5
Vector with dot() : 1.6665e-5
2D Matrix with sum{}: 0.000349687
2D Matrix with bigdot(): 0.000601139
3D Matrix with sum{}: 0.163171205
3D Matrix with dot(): 0.209256164
n = 200
Vector with sum{}: 3.2266e-5
Vector with dot() : 6.8455e-5
2D Matrix with sum{}: 0.001576618
2D Matrix with bigdot(): 0.00817188
3D Matrix with sum{}: 1.318034562
3D Matrix with dot(): 1.260684223
n = 300
Vector with sum{}: 1.7312e-5
Vector with dot() : 3.6582e-5
2D Matrix with sum{}: 0.003396242
2D Matrix with bigdot(): 0.011725772
3D Matrix with sum{}: 8.917606149
3D Matrix with dot(): 12.268530757
n = 400
Vector with sum{}: 3.4863e-5
Vector with dot() : 5.1579e-5
2D Matrix with sum{}: 0.017796441
2D Matrix with bigdot(): 11.815530775
3D Matrix with sum{}: 31.535750208
3D Matrix with dot(): 32.0213833

❯ j test/perf/speed.jl                                                    JuMP/git/O̶n̶e̶I̶n̶d̶e̶x̶e̶d̶A̶r̶r̶a̶y̶ !
PMEDIAN BUILD MIN=0.880151626  MED=0.99301881
PMEDIAN WRITE MIN=2.57723014  MED=2.808084178
CONT5 BUILD   MIN=0.316654603  MED=0.455836884
CONT5 WRITE   MIN=2.004549721  MED=2.113345455

❯ j test/perf/speed2.jl                                                    JuMP/git/O̶n̶e̶I̶n̶d̶e̶x̶e̶d̶A̶r̶r̶a̶y̶
PMEDIAN BUILD MIN=0.609861691  MED=0.811381393
PMEDIAN INTRN MIN=0.591063064  MED=0.77370961
CONT5 BUILD   MIN=0.210495195  MED=0.369529676
CONT5 INTRN   MIN=0.659219582  MED=0.772653258
```
Master:
```
❯ j test/perf/macro.jl                                                              JuMP/git/master
  Running N=20...
    N=20 min 0.014011159
    N=20 min 0.027075327
  Running N=50...
    N=50 min 0.210435025
    N=50 min 0.49117288
  Running N=100...
    N=100 min 1.983629881
    N=100 min 4.033370674

❯ j test/perf/vector_speedtest.jl                                                   JuMP/git/master
n = 10
Vector with sum{}: 8.111e-6
Vector with dot() : 1.005e-5
2D Matrix with sum{}: 5.348e-6
2D Matrix with bigdot(): 0.014400985
3D Matrix with sum{}: 2.9851e-5
3D Matrix with dot(): 6.8458e-5
n = 50
Vector with sum{}: 1.5876e-5
Vector with dot() : 2.1899e-5
2D Matrix with sum{}: 9.8858e-5
2D Matrix with bigdot(): 0.000224624
3D Matrix with sum{}: 0.012139363
3D Matrix with dot(): 0.007423394
n = 100
Vector with sum{}: 1.206e-5
Vector with dot() : 1.256e-5
2D Matrix with sum{}: 0.000304944
2D Matrix with bigdot(): 0.000792367
3D Matrix with sum{}: 0.148381849
3D Matrix with dot(): 0.195990025
n = 200
Vector with sum{}: 1.2128e-5
Vector with dot() : 2.4133e-5
2D Matrix with sum{}: 0.001853661
2D Matrix with bigdot(): 0.004033623
3D Matrix with sum{}: 1.143839756
3D Matrix with dot(): 1.4603047
n = 300
Vector with sum{}: 1.2438e-5
Vector with dot() : 2.7391e-5
2D Matrix with sum{}: 0.003956028
2D Matrix with bigdot(): 0.00859415
3D Matrix with sum{}: 9.505288036
3D Matrix with dot(): 12.334944907
n = 400
Vector with sum{}: 4.4337e-5
Vector with dot() : 7.0052e-5
2D Matrix with sum{}: 0.015749665
2D Matrix with bigdot(): 20.96159887
3D Matrix with sum{}: 33.669532474
3D Matrix with dot(): 52.771684132

❯ j test/perf/speed.jl                                                              JuMP/git/master
PMEDIAN BUILD MIN=0.827251964  MED=0.909822945
PMEDIAN WRITE MIN=2.370976961  MED=2.553967133
CONT5 BUILD   MIN=0.314039771  MED=0.383290275
CONT5 WRITE   MIN=1.982154237  MED=2.067859643

❯ j test/perf/speed2.jl                                                             JuMP/git/master
PMEDIAN BUILD MIN=0.579676763  MED=0.779409968
PMEDIAN INTRN MIN=0.557345187  MED=0.640024864
CONT5 BUILD   MIN=0.202614299  MED=0.35652461
CONT5 INTRN   MIN=0.555429432  MED=0.709362282
```
So some regression on the speed tests, some (sometimes nontrivial) gains on the others. Kind of a mixed bag.

File: src/JuMPContainer.jl (hunk 3 above)
Comment (joehuchette) [inline src/JuMPContainer.jl]: Rewriting this to eval this definition sees a marginal performance increase for iteration (e.g. 17 secs vs 19 secs)

File: src/JuMP.jl [hunk 7]
```diff
@@ -76,6 +76,9 @@ type Model
     colUpper::Vector{Float64}
     colCat::Vector{Symbol}
 
+    # Variable cones of the form, e.g. (:SDP, 1:9)
+    varCones::Vector{Tuple{Symbol,Any}}
```
Comment (mlubin) [inline src/JuMP.jl]: I like this, will make it easy to support new cones.

**Comment by mlubin** [comment]
NEWS entry?

**Pull request merged**